Initial peers tutorial (#117)
eProsima/vulcanexus: docs/rst/tutorials/core/qos/initial_peers/initial_peers.rst

A complete description of the initial peers list and its configuration can be found in Initial peers |InitialPeersFastDdsDocs|.

.. _initial_peers_prerequisites:

Prerequisites
-------------

First of all, make sure that Vulcanexus Humble is installed.
The docker installation is required for this tutorial (see :ref:`Docker installation <docker_installation>`).

Also, complete the :ref:`Configuring Fast-DDS QoS via XML profiles <tutorials_xml_profiles_intro>` tutorial to ensure the execution of the XML profile configuration.

Open two terminals, and run the Vulcanexus Humble image in each one with the following command:

.. code-block:: bash

      docker run \
        -it \
        --name ros2_context_1 \ #rename ros2_context_2 in the second case
        ubuntu-vulcanexus:humble-desktop

.. _initial_peers_add_qos:

Adding initial peers QoS Policy
-------------------------------

In the docker container named ``ros2_context_1``, create a XML file, and complete it with the following example:

.. literalinclude:: /resources/tutorials/core/qos/initial_peers/initial_peers_config.xml
    :language: xml

.. note::

>     For advanced users, and according to the `RTPS standard <https://www.omg.org/spec/DDSI-RTPS/2.2/PDF>`_ (Section 9.6.1.1), the participant discovery traffic unicast listening ports are defined in the `Well Known Ports <https://fast-dds.docs.eprosima.com/en/latest/fastdds/transport/listening_locators.html?highlight=well%20known#well-known-ports>`_ section of the Fast DDS documentation.

Reviewer: ```suggestion
    (Advanced users) According to the `RTPS standard <https://www.omg.org/spec/DDSI-RTPS/2.2/PDF>`_ (Section 9.6.1.1), the participant discovery traffic unicast listening ports are defined in the `Well Known Ports <https://fast-dds.docs.eprosima.com/en/latest/fastdds/transport/listening_locators.html?highlight=well%20known#well-known-ports>`_ section of the Fast DDS documentation.
```

    Thus, in this example, the Participant operates in Domain 0 (default domain) and its ID is 1, so its discovery traffic unicast listening port is 7412.
    By default *eProsima Fast DDS* uses as initial peers the metatraffic multicast locators (network addresses).

>     In the last part of the example it is described how to avoid multicast, using unicast and initial peers simultaneously.

Reviewer: ```suggestion
```

> It is required to include the XML file in at least one of the Vulcanexus docker containers.
> As long as they are using multicast to discover each other, the participant with the initial peer set would let the other participant know the address and port through which they will communicate.

Reviewer: Multicast is just for exchanging the unicast addresses so it should work also without multicast communication. 

```suggestion
It is required to include the XML file in at least one of the Vulcanexus docker containers, as the participant with the initial peer set would let the other participant know the address and port through which they will communicate.
```

| CONTAINER ID    | IMAGE                            | COMMAND        | CREATED       | STATUS       | PORTS | NAMES          |
+=================+==================================+================+===============+==============+=======+================+
| <container ID1> | ubuntu-vulcanexus:humble-desktop | "…vulcanexus…" | 2 minutes ago | Up 2 minutes |       | ros2_context_1 |
+-----------------+----------------------------------+----------------+---------------+--------------+-------+----------------+
| <container ID2> | ubuntu-vulcanexus:humble-desktop | "…vulcanexus…" | 2 minutes ago | Up 2 minutes |       | ros2_context_2 |
+-----------------+----------------------------------+----------------+---------------+--------------+-------+----------------+

Using the container name, check its detailed information.
The following command also filters the output to obtain only the container IP address:

.. code-block:: bash

      docker inspect ros2_context_2 | grep IPAddress

Edit the previously XML file created in the ``ros2_context_1`` container, and include the IP address of the ``ros2_context_2`` container.

> Make sure the IP address set in the xml document is NOT the container's own IP address.

Reviewer: ```suggestion
Make sure the IP address set in the xml file is NOT the container's own IP address.
```

.. _initial_peers_xml_location:

XML configuration file location
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> In order to use the profiles, it is required to let the application know it.

Reviewer: ```suggestion
In order to use the profiles loaded from XML configuration files, it is required to let the application know it.
```

> There are two possibilities for providing *Fast DDS* with XML configuration files:

Reviewer: ```suggestion
There are two possibilities for providing ROS 2 with XML configuration files:
```

    .. code-block:: bash

        ros2 run demo_nodes_cpp talker

* ``ros2_context_2`` container (terminal 2):

    .. code-block:: bash

        ros2 run demo_nodes_cpp listener

The listener would receive the messages sent by the talker.

.. _initial_peers_unicast_example:

Unicast Example
^^^^^^^^^^^^^^^

In the previous example, both multicast and unicast communication were used for entity discovery.
Next, it is shown how to fully disable multicast communication for discovery, so that the only means of discovery is through unicast communication.
In this case, one of the participants will know in advance the listening address of the other participant and will send its discovery information via unicast communication.
The second participant will then reply to the first participant using the network address contained in the information received.
After the discovery, both will start the transmission of user data.

Replace the content of the existing XML configuration file in the ``ros2_context_1`` container with the one shown below.

.. literalinclude:: /resources/tutorials/core/qos/initial_peers/initial_peers_unicast_config_context1.xml
    :language: xml

> In this case, by defining a meta-traffic unicast locator, the Participant creates a unicast meta traffic receiving resource (communication status data and discovery protocol data) for each address-port pair specified (see `Disabling all Multicast Traffic <https://fast-dds.docs.eprosima.com/en/latest/fastdds/transport/disabling_multicast.html?highlight=meta%20traffic#disabling-all-multicast-traffic>`_).

Reviewer: ```suggestion
In this case, by defining a meta-traffic unicast locator, the Participant creates a unicast meta-traffic receiving resource (communication status and discovery protocol data) for each address-port pair specified (see `Disabling all Multicast Traffic <https://fast-dds.docs.eprosima.com/en/latest/fastdds/transport/disabling_multicast.html?highlight=meta%20traffic#disabling-all-multicast-traffic>`_).
```

In that way, and by setting a custom port as initial peer, both ROS 2 context would communicate exclusively by each specified network address and ports (avoiding multicast and the usage of any default ports).

> Create a new XML file and complete it with the following example in the ``ros2_context_2`` container:

Reviewer: ```suggestion
Then, create a new XML file and complete it with the following example in the ``ros2_context_2`` container:
```

.. _initial_peers_xml_location:

XML configuration file location
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

In order to use the profiles loaded from XML configuration files, it is required to let the application know it.
There are two possibilities for providing ROS 2 with XML configuration files:

*   **Recommended**: Setting the location in ``FASTRTPS_DEFAULT_PROFILES_FILE`` environment variable, which may contain the path to the XML configuration file (see `Environment Variables <https://fast-dds.docs.eprosima.com/en/latest/fastdds/env_vars/env_vars.html#env-vars>`_).

    .. code-block:: bash

        export FASTRTPS_DEFAULT_PROFILES_FILE=<path_to_xml_file>

*   **Alternative**: Renaming the XML file to *DEFAULT_FASTRTPS_PROFILES.xml* and placing it in the running application directory.
    For example:

    .. code-block:: bash

        cp <path_to_xml_file> DEFAULT_FASTRTPS_PROFILES.xml
        export RMW_FASTRTPS_USE_QOS_FROM_XML=1

.. _initial_peers_example:

Example
-------

Once the ROS 2 context running in ``ros2_context_1`` Vulcanexus container has been configured, run a demo application in each container.

> * ``ros2_context_1`` container (terminal 1):

Reviewer: ```suggestion
*   ``ros2_context_1`` container (terminal 1):
```

    .. code-block:: bash

        ros2 run demo_nodes_cpp talker

> * ``ros2_context_2`` container (terminal 2):

Reviewer: ```suggestion
*   ``ros2_context_2`` container (terminal 2):
```

.. _initial_peers_ip_address:

Obtain Vulcanexus docker instance IP address
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Open a terminal, and list the Vulcanexus docker instances.

.. code-block:: bash

      docker container list

The output should look like the following table:

+-----------------+----------------------------------+----------------+---------------+--------------+-------+------------+
| CONTAINER ID    | IMAGE                            | COMMAND        | CREATED       | STATUS       | PORTS | NAMES      |
+=================+==================================+================+===============+==============+=======+============+
| <container ID1> | ubuntu-vulcanexus:humble-desktop | "…vulcanexus…" | 2 minutes ago | Up 2 minutes |       | keen_john  |
+-----------------+----------------------------------+----------------+---------------+--------------+-------+------------+
| <container ID2> | ubuntu-vulcanexus:humble-desktop | "…vulcanexus…" | 2 minutes ago | Up 2 minutes |       | eager_pike |
+-----------------+----------------------------------+----------------+---------------+--------------+-------+------------+

> .. important::

Author: Got it!

.. code-block:: bash

      docker inspect <instance_autogenerated_name(keen_jones)> | grep IPAddress

Edit the previously created XML file, and include the IP address of the **other** instance.

.. important::

    Make sure the IP address set in the xml document is NOT the instance's own IP address.

.. _initial_peers_xml_location:

XML configuration file location
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

In order to use the profiles, it is required to let the application know it.
There are two possibilities for providing *Fast DDS* with XML configuration files:

* **Recommended**: Setting the location with environment variable ``FASTRTPS_DEFAULT_PROFILES_FILE`` to contain the path to the XML configuration file (see `Environment Variables <https://fast-dds.docs.eprosima.com/en/latest/fastdds/env_vars/env_vars.html#env-vars>`_).

.. code-block:: bash

      export FASTRTPS_DEFAULT_PROFILES_FILE=<path_to_xml_file>

* **Alternative**: Placing the XML file in the running application directory under the name *DEFAULT_FASTRTPS_PROFILES.xml*.

For example:

.. code-block:: bash

    cp <path_to_xml_file> DEFAULT_FASTRTPS_PROFILES.xml
    export RMW_FASTRTPS_USE_QOS_FROM_XML=1

.. important::

>     Repeat the same procedure in the other opened Vulcanexus docker instance.

Author: You were right, this section was  removed after revision

.. _initial_peers_example:

Example
-------

Once both Vulcanexus docker environments have been configured, run a demo application in each instance:

.. code-block:: bash

      # Talker
      ros2 run demo_nodes_cpp talker

      # Listener

>       ros2 run demo_nodes_cpp listener

Reviewer: *   ``ros2_context_1`` container (terminal 1):

    .. code-block:: bash

        ros2 run demo_nodes_cpp talker

*   ``ros2_context_2`` container  (terminal 2):

    .. code-block:: bash

        ros2 run demo_nodes_cpp listener

> The Vulcanexus docker instances would discover each other via multicast, then one of them would let the other know its address and port through it would establish any communication, and they will connect.

Reviewer: Remove this

.. _initial_peers_unicast_example:

Unicast Example
^^^^^^^^^^^^^^^

> If unicast is required instead of multicast, the following profile should be updated in the existing XML profile:

Reviewer: ```suggestion
In the previous example, both multicast and unicast communication were used for entity discovery. Next, it is shown how to fully disable multicast communication for discovery, so that the only means of discovery is through unicast communication. In this case, one of the participants will know in advance the listening address of the other participant and will send its discovery information via unicast communication. The second participant will then reply to the first participant using the network address contained in the information received. After the discovery, both will start the transmission of user data. 

Replace the content of the existing XML configuration file with the one shown below.
```

> .. code-block:: xml

Reviewer: Save both files in the resources folder.

    <?xml version="1.0" encoding="UTF-8" ?>
        <profiles xmlns="http://www.eprosima.com/XMLSchemas/fastRTPS_Profiles">
            <participant profile_name="initial_peers_example_profile" is_default_profile="true">
                <rtps>
                    <builtin>
                        <initialPeersList>
                            <locator>
                                <udpv4>
                                    <address>172.17.0.3</address> <!-- "keen_john"  IP address -->

>                                     <port>7777</port>             <!-- "keen_john"  port       -->

Reviewer: Update container names

                                </udpv4>
                            </locator>
                        </initialPeersList>
                        <metatrafficUnicastLocatorList>
                            <locator>
                                <udpv4>
                                    <address>172.17.0.2</address> <!-- "eager_pike" IP address -->

>                                     <port>7666</port>             <!-- "eager_pike" port       -->

Reviewer: Update container names

                                </udpv4>
                            </locator>
                        </metatrafficUnicastLocatorList>
                    </builtin>
                </rtps>
            </participant>
        </profiles>

In this case, the port `7777` is not longer the default port, it is a new custom port to reach the peer address.

> Furthermore, the meta-traffic unicast locator has been set with the Vulcanexus docker instance IP address (own address), and the port it would be communicating through.

Reviewer: Comment first what the ``metatrafficUnicastLocatorList`` does. Then explain that the same will apply to the other so to reach it out we should change the initial peer configuracion.

> Create a new XML file and complete it with the following example in the non-existing xml file docker instance:

Reviewer: Refer to the second container by it name (``ros2_context_2``).

.. literalinclude:: /resources/tutorials/core/qos/initial_peers/initial_peers_unicast_config.xml
    :language: xml

> Finally, the addresses and ports have been swapped compared to the previous XML file, in order to match IP addresses and ports in which they are set to use.

Reviewer: Remove or rephrase as it's difficult to understand what you want to say with this swap.

> .. important::

Reviewer: Not important, just the continuation of the tutorial.

    Make sure the XML profile has been configured in both docker instances:

    .. code-block:: bash

      export FASTRTPS_DEFAULT_PROFILES_FILE=<path_to_xml_file>

Run the application example again in both Vulcanexus docker instances:

.. code-block:: bash

      # Talker
      ros2 run demo_nodes_cpp talker

      # Listener

>       ros2 run demo_nodes_cpp listener

Reviewer: Same comment as before. Update with the list and name of the containers
